jax2onnx/converter/ir_optimizations.py

```diff
@@ -0,0 +1,1741 @@
+        outs = getattr(graph, attr, None)
+        if outs is None:
+            continue
+        try:
+            lst = list(outs)
+        except Exception:
+            continue
+        changed = False
+        for i, ov in enumerate(lst):
+            if (old_v is not None and ov is old_v) or (
+                old_name and _v_name(ov) == old_name
+            ):
+                lst[i] = new_v
+                changed = True
+        if changed:
+            try:
+                setattr(graph, attr, lst)
+            except Exception:
+                # last resort: rename
+                for i, ov in enumerate(getattr(graph, attr, [])):
+                    if (old_v is not None and ov is old_v) or (
+                        old_name and _v_name(ov) == old_name
+                    ):
+                        try:
+                            ov.name = _v_name(new_v)
+                        except Exception:
+                            pass
+
+
+def _build_use_maps(nodes: List["ir.Node"]):
+    prod_by_obj: Dict[int, int] = {}
+    prod_by_name: Dict[str, int] = {}
+    cons_by_obj: Dict[int, Set[int]] = defaultdict(set)
+    cons_by_name: Dict[str, Set[int]] = defaultdict(set)
+    for i, n in enumerate(nodes):
+        for ov in _node_outputs(n):
+            if ov is None:
+                continue
+            prod_by_obj[id(ov)] = i
+            nm = _v_name(ov)
+            if nm:
+                prod_by_name[nm] = i
+    for i, n in enumerate(nodes):
+        for iv in _node_inputs(n):
+            if iv is None:
+                continue
+            cons_by_obj[id(iv)].add(i)
+            nm = _v_name(iv)
+            if nm:
+                cons_by_name[nm].add(i)
+    return prod_by_obj, prod_by_name, cons_by_obj, cons_by_name
+
+
+def _unique_consumer(
+    cons_by_obj, cons_by_name, val: Optional["ir.Value"]
+) -> Optional[int]:
+    if val is None:
+        return None
+    nm = _v_name(val)
+    S = set(cons_by_obj.get(id(val), set()))
+    if nm:
+        S |= set(cons_by_name.get(nm, set()))
+    return next(iter(S)) if len(S) == 1 else None
+
+
+def _producer_idx_for(val: Optional["ir.Value"], prod_obj, prod_name) -> Optional[int]:
+    if val is None:
+        return None
+    nm = _v_name(val)
+    return prod_obj.get(id(val)) or (prod_name.get(nm) if nm else None)
+
+
+def _all_consumers(cons_by_obj, cons_by_name, v: Optional["ir.Value"]) -> Set[int]:
+    if v is None:
+        return set()
+    nm = _v_name(v)
+    S = set(cons_by_obj.get(id(v), set()))
+    if nm:
+        S |= set(cons_by_name.get(nm, set()))
+    return S
+
+
+# ---------------- Attr access ----------------
+
+
+def _get_attr(node, name: str):
+    """
+    Robustly fetch a node attribute 'name' across onnx_ir variants:
+      • dict-like or mapping containers (Attributes)
+      • sequences of Attrs
+      • direct attribute on the node
+    """
+    for attr_name in ("attributes", "attrs", "_attrs"):
```

Comment:
This is redundant. You may refer to the api docs for efficiently retriving the values https://onnx.ai/ir-py/api/generated/onnx_ir._graph_containers.Attributes.html#onnx_ir._graph_containers.Attributes